Four new SparseMatrix methods, enabling batch synapse learning

mrcslws opened the pull request:
Fixes #1137 
Fixes #1138 

Changes:

- Expose the `(nrows, ncols)` SparseBinaryMatrix constructor in the bindings
- Add four methods to the SparseMatrix. These methods can be used do typical synapse learning in batches.

The SparseMatrix tests are in nupic. Maybe they should be moved to nupic.core. For now, I'm putting tests for each of these methods in nupic.

mrcslws commented:
Here's a before-and-after.

Before:
https://github.com/numenta/nupic.research/blob/abac7c829454eaa4781f6babf4d6779ce687205e/htmresearch/algorithms/column_pooler.py#L447

After: https://github.com/mrcslws/nupic.research/blob/c2e6fe9199919c12c9ec9baf6e12ce5ac37f842e/htmresearch/algorithms/column_pooler.py#L443

The same logic is expressed in these four method calls. And it's fast.

(Technically we weren't clipping the values before, so I also do that. The "Before" code can be implemented with three method calls.)

mrcslws commented:
@subutai @scottpurdy This is ready for review. I think both of you will at least want to glance at this.

Note the usage example in nupic.research and the tests in nupic.

rhyolight commented:
test

scottpurdy commented:
Are these new four methods generally useful or specific to the column pooler or other HTM-related algorithms?

mrcslws commented:
I aimed for "generally useful". They're pretty generic. They'd be useful in lots of synapse learning algorithms. And they're designed to feel like the other SparseMatrix methods, e.g. using "Outer" the same way.

scottpurdy commented:
Re: testing, there should be C++ tests added for your functions. And the Python unit tests should be added in this repo (in `bindings/py/tests`) rather than in nupic.

mrcslws commented:
@scottpurdy @oxtopus I moved the nupic.bindings math tests from nupic. I spent some time figuring out how to preserve file history, even to before the previous rename (which is extra difficult). I manually listed the previous filenames, then I used this approach:

~~~bash
cd ~/nta/nupic

git log --pretty=email --patch-with-stat --reverse --full-index --binary -- \
tests/unit/py2/nupic/math/array_algorithms_test.py \
tests/unit/py2/nupic/math/cast_mode_test.py \
tests/unit/py2/nupic/math/lgamma_test.py \
tests/unit/py2/nupic/math/nupic_random_test.py \
tests/unit/py2/nupic/math/sparse_binary_matrix_test.py \
tests/unit/py2/nupic/math/sparse_matrix_test.py \
tests/unit/nupic/math/array_algorithms_test.py \
tests/unit/nupic/math/cast_mode_test.py \
tests/unit/nupic/math/lgamma_test.py \
tests/unit/nupic/math/nupic_random_test.py \
tests/unit/nupic/math/sparse_binary_matrix_test.py \
tests/unit/nupic/math/sparse_matrix_test.py \
> ~/Desktop/everything.patch

cd ~/nta/nupic.core

git am < ~/Desktop/everything.patch
~~~

mrcslws commented:
@scottpurdy @subutai I made the following changes:

- Moved the nupic.bindings math tests to nupic.core, preserving change history.
- Added detailed docstrings for each method.
- Wrote matching C++ unit tests. They're basically the same tests as the Python tests.
- Added two additional SparseMatrix methods to support the new 'sampleSize' parameter.
